src/main/java/org/apache/sling/scripting/core/impl/ScriptingVariablesNonceFactory.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2016 The Apache Software Foundation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.sling.scripting.core.impl;
+
+import java.security.SecureRandom;
+
+import org.apache.commons.collections4.map.LRUMap;
+
+/**
+ * Factory for generating one-time-use tokens for the 
+ * ScriptingVariablesConsolePlugin + SlingBindingsVariablesListJsonServlet
+ * transactions.
+ */
+final class ScriptingVariablesNonceFactory {
+    private static final String EMPTY_STRING = "";
+	private static final SecureRandom secureRandom = new SecureRandom();
+    // use LRUMap so it is not possible to grow indefinitely if isValid is
+    //  never called after nextNonce is called
+    private static final LRUMap<String, String> validNonceSet = new LRUMap<>(100);
+
+    /**
+     * Generate a random use once value
+     * 
+     * @return random nonce value
+     */
+    public static String nextNonce() {
+        byte[] nonce = new byte[8];
+        secureRandom.nextBytes(nonce);
+        String nonceHex = convertBytesToHex(nonce);
+        validNonceSet.put(nonceHex, EMPTY_STRING);
+        return nonceHex;
+    }
+
+    /**
+     * Check if the supplied value matches one of the
+     * previously generated single use values
+     * 
+     * @return true if the nonce was valid, false otherwise
+     */
+    public static boolean isValid(String nonce) {
+        Object remove = validNonceSet.remove(nonce);
```

Comment:
Would it make sense to validate with the help of a SHA256 of the nonce being added in addition to the nonce to the request? That would allow to not store the nonces on server side

Reply:
I'm open to a solution that doesn't require any server side state, but I am unclear what exactly I would be hashing that I could compare on next request that couldn't be easily spoofed.

Did you have something specific in mind?

Reply:
I would recommend to use two request parameters: nonce and nonceHmacSha256. This is valid and cannot be easily spoofed as the secret used for the HmacSha256 is not exposed. You could even prevent replay attacks by using the current server time as nonce and only accept it for 5 minutes.

Reply:
For performance reasons we could also consider using the less secure HmacSha1 (https://docs.oracle.com/javase/8/docs/api/javax/crypto/Mac.html).

Reply:
Ok, I see what you are suggesting.  That should work good enough.

I've made the changes for review.